File: leetcode/java/validPalindrome.java
/*
 * valid a string whether is palindrome, considering only alphanumeric char and ignoring cases.
 *
 * NOTE: during interview, it is good to ask what about empty string? here we consider empty string is palindrome
 * */
import java.util.*;
import java.io.*;

public class validPalindrome{
    public boolean isPalindrome(String s){
        if(s.isEmpty())    return true;
        int i=0, j=s.length()-1;
        while(i<j){
            if(!Character.isLetterOrDigit(s.charAt(i))){    //static isLetterOrDigit()
                ++i;
                continue;
            }
            if(!Character.isLetterOrDigit(s.charAt(j))){
                --j;
                continue;
            }
            Character a = new Character(s.charAt(i));    //create one Charactder object to call equals()
            if(Character.isLetter(a))    a = Character.toLowerCase(a);
            char b = s.charAt(j);
            if(Character.isLetter(b))    b = Character.toLowerCase(b);
            if(!a.equals(b))    break;            //equals() is not static
            ++i;
            --j;
        }
        return i>=j;
    }

    public static void main(String[] args){
        Scanner scan = new Scanner(System.in);
        validPalindrome vp = new validPalindrome();
        while(true){
            System.out.println("please input string to valid:");
            String str = scan.nextLine();
            System.out.println(vp.isPalindrome(str) ? "true" : "false");
        }
    }
}

File: leetcode/src/validPalindrome.cpp
/*
 * given a string, determine if it is a palindrome, consider only alphanumeric char and ignoring cases.
 *
 * NOTE: it is good to ask during interview, what about empty string? here we determine empty is true
 *
 * */
#include <string>
#include <cctype>

using namespace std;

class Solution{

public:
    bool isPalindrome(const string& s){
        if(s.empty())    return true;

        int i = 0, j = s.size() - 1;

        while(i < j){

            if(!isalpha(s[i]) && !isdigit(s[i])){
                ++i;
                continue;
            }

            if(!isalpha(s[j]) && !isdigit(s[j])){
                --j;
                continue;
            }

            char a = isalpha(s[i]) ? tolower(s[i]) : s[i];
            char b = isalpha(s[j]) ? tolower(s[j]) : s[j];

            if(a != b)    break;

            ++i;
            --j;
        }

        return i >= j;
    }
};

/* unit test is in ../cpp_unittest/validPalindrome_unittest */
